### ISSUE TITLE
Multiple release channels in one repository

### ISSUE THREAD

**Issue by jsj1027** [flatpak/flatpak#4192]
I know flathub/flatpaks does technically have beta channels, but in my opinion right now they don't make a great UX. 

 From my understanding to install a Flatpak from a beta channel you have to do the following,
 
### Install Beta Flatpaks

1. Add the flathub beta repo _flatpak remote-add flathub-beta https://flathub.org/beta-repo/flathub-beta.flatpakrepo_
2. Install the app using the repo beta option _flatpak install --user flathub-beta org.godotengine.Godot_
3. And then you can run the beta app using _flatpak run org.godotengine.Godot//beta_

Now the great thing is beta and stable channel apps can be installed in parallel. However it really feels like bad ux in comparison to Snaps.

### Install Beta Snaps

1. Install the beta snap_sudo snap install skype --channel=insider_
2. Then you can run it.

However it should be noted that you can not install a beta snap and a stable snap of the same package on the same machine.

### Where to go from here

I would love to see a way to install a beta flatpak in a much easier way. Something like _flatpak install --user flathub org.godotengine.Godot --channel beta_

Simple clean, no need to add a beta repo, and these beta channels can be advertised on Flathub itself, just like snaps can show they have beta/insider/other release channels right on Snapcraft.

**Comment by smcv** [comment]
I've retitled this to "Multiple release channels in one repository" because I think that's a better description of the feature you're asking for.

Unlike Snap, Flatpak is designed without assuming that there is a single, central, canonical (or Canonical) repository: you can add zero or more remotes, and install any mixture of apps from those remotes. As such, changing design decisions made in Flatpak is a Flatpak feature request, but changing design decisions made in Flathub (the main repository that we encourage people to use) is really a request for Flathub rather than Flatpak.

Flathub *could* put all their packages in one big repository, if they wanted to: it would be technically completely possible to make `flatpak install flathub org.godotengine.Godot//beta` do what you think it would do. That seems like the closest equivalent to how you say Snap does this.

However, the people running Flathub have chosen not to do that. I am not them, so I don't know whether their reasons are good or bad.

Some advantages to the beta repository being separate:

* If an app is not really ready yet, it would make sense to *only* make it available as a beta, with no stable version existing. We did this when https://github.com/flathub/com.valvesoftware.SteamLink was first published. If the app isn't really ready, we don't necessarily want `flatpak install flathub com.example.App` to accidentally install the beta - we'd prefer people to only get betas or other prereleases if they specifically ask for them.
* If a user is not interested in using beta or prerelease software, they don't have to waste their bandwidth on downloading the metadata of the betas.
* The Flathub instructions nudge users towards installing the stable version system-wide, and the beta version in their home directory. That seems like it makes a lot of sense for a multi-user system: if I share a system with someone with a higher tolerance for regressions, I don't want to be accidentally running the beta version of an app I rely on, just because the other user installed it for their own use.

**Comment by jsj1027** [comment]
I appreciate the name change, I feel "Multiple release channels in one repository" really does make it more clear.

So to respond I do agree with your points as these are some great technical pros. However I feel like there are some UX cons that arent being seen.

To get us on the same page lets figure hacking out a starting point of assumptions.

### Assumptions

- Flatpaks are found and installed by remote name and then full org+application name
   -  `flatpak install flathub com.example.App` installs app probably the stable version
   -  `flatpak install flathub com.example.App//beta` installs app from the same repo, however its the apps beta version from same repo. Not because it is on a specific beta repo but because the name includes `//beta`
- Flatpak grabs the meta data for the entire repo upon adding a repo, and updates this metadata periodically
- `flatpak install flathub com.example.App` can install the beta release of an app

If and only if all this is true and correct I fully agree with you. However what if we could make things work a bit differently? 

### Future State(? Im bad with titles)

-  Flatpaks are found and installed by remote name, full org+application name and then channel
   - `flatpak install flathub com.example.App` defaults to stable
   - The install arg has a optional arguments called `--channel`. Which installs the beta version of that app from that specific repo. Keeping the ability to
      - > you can add zero or more remotes, and install any mixture of apps from those remotes.
   -  Alpha/Beta/Insiders/Etc are optional for an app to have
- Flatpak grabs the meta data for the entire repo upon adding a repo, and updates this metadata periodically
  - Now the metadata includes a list of channels available for a package, rather than adding the full metadata for a new package. This should lead to less metadata in a case where someone does want beta channels. And it should hopefully decrease metadata concerns.
 - `flatpak install flathub com.example.App` will **Never** install the beta release of an app
    - This defaults to stable and if an app only exists on non stable channels the user can be given an error message and be told what channels the app does actually have.
 - Flatpak's can also default to installing non stable versions of a package to a user's home dir instead of system wide.

**Comment by jsj1027** [comment]
Any thoughts on this @smcv?

**Comment by cassidyjames** [comment]
As another data point, we were considering using branches for beta/daily versions of elementary and AppCenter apps alongside stable releases in our Flatpak repositories, but @jsj1027's last assumption is the reason we cannot for now; if an app has _only_ been pushed only as a daily or beta version (e.g. before it has been approved for release), then `flatpak install flathub com.example.App` cannot install it without an additional explicit intent to install the daily or unstable version.

So, we'll just use separate repositories much like Flathub for the beta remote. A potential solution for us would be for a repository to be able to have a `default` branch, and for `flatpak install` to not automatically suggest/install apps from a different branch than the default without explicitly including the branch name e.g.with `flatpak install flathub com.example.App//beta`.

**Comment by smcv** [comment]
> A potential solution for us would be for a repository to be able to have a default branch, and for flatpak install to not automatically suggest/install apps from a different branch than the default without explicitly including the branch name

So what's the backwards compatibility story there? Users of old versions of `flatpak install` accidentally get the betas anyway? That doesn't seem good.

Would it work if there was a way for the repository metadata to say "prerelease versions can be fetched from this other repository", in a way that was only understood by new Flatpak? Then old versions would require specific user action (like Flathub has now), but new versions would be free to interpret `flatpak install myrepo com.example.App//beta` as: if `com.example.App//beta` exists in `myrepo`, use it, else if it exists in the lookaside repository mentioned in `myrepo`'s metadata, use that?

**Comment by smcv** [comment]
I don't think inventing a new argument like `--channel` is a good answer to this, because "having several release channels" is exactly what branches are for.

**Comment by samtuke** [comment]
I think I'm missing this functionality. Use-case: trying to install the latest version of Thunderbird. It's available on Snapcraft as a release candidate from the standard source / repo using the command `sudo snap install thunderbird --candidate`. Seems like flatpak offers nothing in this vein for me as a user or Thubderbird as a publisher / vendor.

**Comment by cassidyjames** [comment]
@samtuke Flatpak offer different remotes, and Flathub maintains a beta remote. So, you could opt into using the beta remote and install an app from there. E.g.:

```
flatpak remote-add --user --if-not-exists beta https://flathub.org/beta-repo/flathub-beta.flatpakrepo
```

Then `flatpak install beta thunderbird` will get you to the right place. 😄

It might be nice to be able to explicitly support non-default channels within the same remote, but it turns out in practice, the above works alright.

**Comment by ninja-** [comment]
I had a look. The UX after manually installing the beta flatpak repo looks OK, and with the new "Software" app center on Ubuntu - it matches the experience of snaps where beta channel is automatically offered as a choice.
Unfortunately, for example latest Chromium beta flatpak was updated 3 years ago - I am not sure of an app that's actually properly maintained in beta channel

**Comment by ninja-** [comment]
![image](https://github.com/flatpak/flatpak/assets/5637961/d12efbaf-63b5-47e2-af6e-871349c51df2)

**Comment by smcv** [comment]
> Unfortunately, for example latest Chromium beta flatpak was updated 3 years ago - I am not sure of an app that's actually properly maintained in beta channel.

The maintainers of Flatpak do not maintain individual apps, or Flathub in general. As I said above, Flatpak is a decentralized system. Flathub is the most popular and widely-recommended Flatpak repository, but it is not the only Flatpak repository that can exist.

I suspect that the reason for Chromium betas not being updated frequently is "packaging a web browser is a massive amount of work". If you can help to maintain an app that is of interest to you, please do: each app that is available on Flathub has its own issue tracking, for example Chromium is at <https://github.com/flathub/org.chromium.Chromium>.

The beta channel on Flathub is actively used for many apps (I know it's definitely used for Steam and Steam Link, for example). The encouraged workflow is to upload every new release as a beta, then test it, then promote it to stable.

> Since snap enables [a beta channel] by default, I believe flatpak should too in new Ubuntu versions, for example.

The maintainers of Flatpak do not control what Ubuntu chooses to do, and do not maintain any particular user interface layer above Flatpak. It would be possible for a user interface implementation to enable a beta channel by default, if its maintainers wanted it to.

**Comment by smcv** [comment]
If someone wants to implement what I suggested in https://github.com/flatpak/flatpak/issues/4192#issuecomment-821377110, they can try it and send a pull request.

I don't think any of the maintainers of Flatpak are particularly interested in working on that (we all have many responsibilities outside Flatpak), so I'm going to close this as "not planned". That doesn't mean that I wouldn't be willing to consider a PR, but realistically, then only way this is going to happen is if one of the people who particularly wants this feature can step up to do the work.

**Comment by bam80** [comment]
> Unfortunately, for example latest Chromium beta flatpak was updated 3 years ago - I am not sure of an app that's actually properly maintained in beta channel.

Organic Maps seem updated from time to time:
https://github.com/flathub/app.organicmaps.desktop/tree/beta
@ferenc-

**Comment by hfiguiere** [comment]
Updates are done by individual application maintainers. This is not the venue to request it.
